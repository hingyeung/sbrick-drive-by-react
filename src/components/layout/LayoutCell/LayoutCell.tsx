import * as React from 'react';
import * as classNames from 'classnames';

export interface Props {
  extraClassNames: string;
  sm?: number;
  md?: number;
  lg?: number;
  xl?: number;
  children: any;
}

function isValidBootstrapColumnValue(value: number) {
  return value >= 1 && value <= 12;
}

function buildBootstrapGridColumnClasses(props: Props): any {
  return {
    [`col-sm-${props.sm}`]: props.sm && isValidBootstrapColumnValue(props.sm),
    [`col-md-${props.md}`]: props.md && isValidBootstrapColumnValue(props.md),
    [`col-lg-${props.lg}`]: props.lg && isValidBootstrapColumnValue(props.lg),
    [`col-xl-${props.xl}`]: props.xl && isValidBootstrapColumnValue(props.xl)
  };
}

const LayoutCell: React.SFC<Props> = (props: Props) => {
  return (
    <div className={classNames(props.extraClassNames, buildBootstrapGridColumnClasses(props))}>
      {props.children}
    </div>
  );
};

export default LayoutCell;