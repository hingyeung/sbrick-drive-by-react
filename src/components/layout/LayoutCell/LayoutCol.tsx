import * as React from 'react';
import * as classNames from 'classnames';
import { StatelessComponent } from 'react';

interface LayoutCellComponentProps {
  className?: string;
  componentClass?: any;
  sm?: number;
  md?: number;
  lg?: number;
  xl?: number;
}

const isValidBootstrapColumnValue = function(value: number) {
  return value >= 1 && value <= 12;
};

const buildBootstrapGridColumnClasses = function(props: any): any {
  return {
    [`col-sm-${props.sm}`]: props.sm && isValidBootstrapColumnValue(props.sm),
    [`col-md-${props.md}`]: props.md && isValidBootstrapColumnValue(props.md),
    [`col-lg-${props.lg}`]: props.lg && isValidBootstrapColumnValue(props.lg),
    [`col-xl-${props.xl}`]: props.xl && isValidBootstrapColumnValue(props.xl)
  };
};

const LayoutCol: StatelessComponent<LayoutCellComponentProps> = (props) => {
  const {
    componentClass: Component,
    className,
    ...otherProps
  } = props;

  return (
    <Component {...otherProps} className={classNames(className, 'col', buildBootstrapGridColumnClasses(otherProps))}/>
  );
};

LayoutCol.defaultProps = {
  componentClass: 'div'
};

export default LayoutCol;