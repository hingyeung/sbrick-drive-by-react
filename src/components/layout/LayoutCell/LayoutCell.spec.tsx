import LayoutCell from './LayoutCell';
import { shallow, mount } from 'enzyme';
import * as React from 'react';

describe('LayoutCell', () => {
  const sizes = ['sm', 'md', 'lg', 'xl'];
  it('should render extraClassNames in props', () => {
    const wrapper = shallow(<LayoutCell extraClassNames="hello">Hello</LayoutCell>);

    expect(wrapper.find('.hello')).toHaveLength(1);
    expect(wrapper).toMatchSnapshot();
  });

  sizes.forEach(size =>
    [1, 12].forEach(width =>
      it(`should set ${size} screen column number when ${size} prop is set`, () => {
        const sizeProps = {[size]: width};
        const wrapper = shallow(
          <LayoutCell extraClassNames="hello" {...sizeProps}>Hello</LayoutCell>
        );

        expect(wrapper.find('.hello').hasClass(`col-${size}-${width}`)).toBeTruthy();
      })
    )
  );

  sizes.forEach(size =>
    [0, 13].forEach(width =>
      it('should not set ${size} screen column number when ${size} prop out of range', () => {
        const sizeProps = {[size]: width};
        const wrapper = mount(
          <LayoutCell extraClassNames="hello" {...sizeProps}>Hello</LayoutCell>
        );

        hasClassWithPrefix(wrapper, `col-${size}-`);
      })
    )
  );

  // it('should set medium screen column number when md prop is set', () => {
  //   const wrapper = shallow(
  //     <LayoutCell extraClassNames="hello" md={6}>Hello</LayoutCell>
  //   );
  //
  //   expect(wrapper.find('.hello').hasClass('col-md-6')).toBeTruthy();
  // });
  //
  // it('should not set medium screen column number when md prop out of range', () => {
  //   const wrapper = mount(
  //     <LayoutCell extraClassNames="hello" md={13}>Hello</LayoutCell>
  //   );
  //
  //   hasClassWithPrefix(wrapper, 'col-md-');
  // });

  const hasClassWithPrefix = (wrapper: any, prefix: string) => {
    expect(
      wrapper.find('.hello').prop('className').trim().split(/\s+/).filter((thisClass: string) =>
        thisClass.indexOf(prefix) > -1)
    ).toEqual([]);
  };
});