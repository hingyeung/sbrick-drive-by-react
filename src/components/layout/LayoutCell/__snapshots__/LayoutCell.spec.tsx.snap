// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LayoutCell should render extraClassNames in props 1`] = `
<div
  className="hello"
>
  Hello
</div>
`;
